File: 백준/Gold/1765. 닭싸움 팀 정하기/닭싸움 팀 정하기.py
import sys

input = sys.stdin.readline


def union(x, y):
    x = find(x)
    y = find(y)

    if x == y:
        return


    if abs(relation[x]) >= abs(relation[y]):
        relation[x] += relation[y]
        relation[y] = x
    else:
        relation[y] += relation[x]
        relation[x] = y


def find(x):
    if relation[x] < 0:
        return x

    relation[x] = find(relation[x])

    return relation[x]


n = int(input())
m = int(input())
graph = [[] for _ in range(n + 1)]
relation = [-1] * (n + 1)

for _ in range(m):
    rel, p, q = input().split()
    p = int(p)
    q = int(q)
    if rel == "E":
        graph[p].append(q)
        graph[q].append(p)
    else:
        union(p, q)

for i in range(1, n+1):
    if (len(graph[i]) > 1):
        for j in range(1, len(graph[i])):
            union(graph[i][j-1], graph[i][j])

answer = 0
for i in range(1, n+1):
    if relation[i] < 0:
        answer += 1

print(answer)

File: 백준/Gold/1765. 닭싸움 팀 정하기/닭싸움 팀 정하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
    static int n, m, p, q;
    static String rel;
    static List<Integer>[] graph;
    static int[] relation;
    public static void union(int x, int y) {
        x = find(x);
        y = find(y);

        if (x == y) return;

        if (Math.abs(relation[x]) >= Math.abs(relation[y])) {
            relation[x] += relation[y];
            relation[y] = x;
        } else {
            relation[y] += relation[x];
            relation[x] = y;
        }
    }
    public static int find(int x) {
        if (relation[x] < 0) {
            return x;
        }

        x = find(relation[x]);

        return x;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        m = Integer.parseInt(br.readLine());
        graph = new List[n + 1];
        relation = new int[n + 1];

        Arrays.fill(relation, -1);

        for (int i = 0; i < n+1; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            rel = st.nextToken();
            p = Integer.parseInt(st.nextToken());
            q = Integer.parseInt(st.nextToken());

            if (rel.equals("F")) {
                union(p, q);
            } else {
                graph[p].add(q);
                graph[q].add(p);
            }
        }

        for (int i = 1; i < n+1 ; i++) {
            if (graph[i].size() > 1) {
                for (int j = 1; j < graph[i].size(); j++) {
                    union(graph[i].get(j-1), graph[i].get(j));
                }
            }
        }

        int ans = 0;
        for (int i = 1; i < n+1; i++) {
            if (relation[i] < 0) {
                ans++;
            }
        }

        System.out.println(ans);
    }
}